packages/e2e-tests/test-utils/event-proxy-server.ts

```diff
@@ -0,0 +1,220 @@
+        sentryResponse => {
+          sentryResponse.addListener('data', (chunk: Buffer) => {
+            proxyResponse.write(chunk, 'binary');
+            sentryResponseChunks.push(chunk);
+          });
+
+          sentryResponse.addListener('end', () => {
+            eventCallbackListeners.forEach(listener => {
+              const rawProxyRequestBody = Buffer.concat(proxyRequestChunks).toString();
+              const rawSentryResponseBody = Buffer.concat(sentryResponseChunks).toString();
+
+              const data: SentryRequestCallbackData = {
+                envelope: parseEnvelope(rawProxyRequestBody, new TextEncoder(), new TextDecoder()),
+                rawProxyRequestBody,
+                rawSentryResponseBody,
+                sentryResponseStatusCode: sentryResponse.statusCode,
+              };
+
+              listener(Buffer.from(JSON.stringify(data)).toString('base64'));
+            });
+            proxyResponse.end();
+          });
+
+          sentryResponse.addListener('error', err => {
+            throw err;
+          });
+
+          proxyResponse.writeHead(sentryResponse.statusCode || 500, sentryResponse.headers);
+        },
+      );
+
+      sentryRequest.write(Buffer.concat(proxyRequestChunks), 'binary');
+      sentryRequest.end();
+    });
+  });
+
+  const proxyServerStartupPromise = new Promise<void>(resolve => {
+    proxyServer.listen(options.port, () => {
+      resolve();
+    });
+  });
+
+  const eventCallbackServer = http.createServer((eventCallbackRequest, eventCallbackResponse) => {
+    eventCallbackResponse.statusCode = 200;
+    eventCallbackResponse.setHeader('connection', 'keep-alive');
+
+    const callbackListener = (data: string): void => {
+      eventCallbackResponse.write(data.concat('\n'), 'utf8');
+    };
+
+    eventCallbackListeners.add(callbackListener);
+
+    eventCallbackRequest.on('close', () => {
+      eventCallbackListeners.delete(callbackListener);
+    });
+
+    eventCallbackRequest.on('error', () => {
+      eventCallbackListeners.delete(callbackListener);
+    });
+  });
+
+  const eventCallbackServerStartupPromise = new Promise<void>(resolve => {
+    const listener = eventCallbackServer.listen(0, () => {
+      const port = String((listener.address() as AddressInfo).port);
+      const tmpFileWithPort = path.join(os.tmpdir(), `event-proxy-server-${options.proxyServerName}`);
+      fs.writeFileSync(tmpFileWithPort, port, { encoding: 'utf8' });
+      resolve();
+    });
+  });
+
+  await eventCallbackServerStartupPromise;
+  await proxyServerStartupPromise;
+  return;
+}
+
+export function waitForRequest(
+  proxyServerName: string,
+  callback: (eventData: SentryRequestCallbackData) => boolean,
+): Promise<SentryRequestCallbackData> {
+  const tmpFileWithPort = path.join(os.tmpdir(), `event-proxy-server-${proxyServerName}`);
+  const eventCallbackServerPort = fs.readFileSync(tmpFileWithPort, 'utf8');
```

Comment:
l: Can this be async?

Reply:
👍 9090f797dcbf9986c216b6ee464cc7aa628badcc